File: 30_days_of_coding/Day_28_RegExPatternsDatabases.cs
class Solution
{
    public static void Main(string[] args)
    {
        int N = Convert.ToInt32(Console.ReadLine().Trim());
        List<string> names = new List<string>();
        string mail = @".+@gmail\.com$";//. - any single char, + matches the previous element 1 or more times, \. - to match a dot not a single char., $ end
        for (int NItr = 0; NItr < N; NItr++)
        {
            string[] firstMultipleInput = Console.ReadLine().TrimEnd().Split(' ');

            string firstName = firstMultipleInput[0];

            string emailID = firstMultipleInput[1];
            
            if(Regex.IsMatch(emailID, mail)){
                names.Add(firstName);
            }
        }
        
        names.Sort();
        
        foreach(var name in names){
            Console.WriteLine(name);
        }
    }
}

File: 30DaysOfCodingJAVA/Day_28_RegExPatternsDatabases.java
public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(bufferedReader.readLine().trim());
        List<String> names = new LinkedList<String>();
        String mail = ".+@gmail\\.com$";
        IntStream.range(0, N).forEach(NItr -> {
            try {
                String[] firstMultipleInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

                String firstName = firstMultipleInput[0];

                String emailID = firstMultipleInput[1];
                
                if(Pattern.matches(mail, emailID)){
                names.add(firstName);
                }
                
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        });
        
        
        
        Collections.sort(names);
        
        for(String name : names){
            System.out.println(name);
        }

        bufferedReader.close();
    }
}
